src/ir/proto_to_ir.cc

```diff
@@ -0,0 +1,260 @@
+    ToIR(block_it.id(), block_it.block());
+  }
+}
+
+void ProtoToIR::ToIR(ProtoToIR::ID block_id, const proto::Block& block_proto) {
+  auto find_block_ptr = blocks_.find(block_id);
+  CHECK(find_block_ptr != blocks_.end())
+      << "This block should be pre-constructed.";
+  auto& block_builder = find_block_ptr->second;
+  // TODO: Consider preserving inputs/outputs.
+  for (const auto& result_it : block_proto.results().values()) {
+    block_builder.AddResult(result_it.first, ToIR(result_it.second));
+  }
+  // Also construct the block's children in place.
+  for (const auto& operation_it : block_proto.operations()) {
+    const auto& operation_id = operation_it.id();
+    const auto& operation_proto = operation_it.operation();
+    auto operation_ir_it = operations_.find(operation_id);
+    CHECK(operation_ir_it != operations_.end())
+        << "Operation " << operation_id << "("
+        << operation_proto.operator_name() << ") was not registered.";
+
+    Operation* operation = operation_ir_it->second.get();
+    CHECK(operation != nullptr) << "Operation ptr is empty!";
+    ToIR(*operation, operation_proto);
+  }
+}
+
+void ProtoToIR::ToIR(Operation& operation,
+                     const proto::Operation& operation_proto) {
+  // TODO: Consider preserving results and impl_module.
+  for (const auto& input_proto : operation_proto.inputs()) {
+    operation.inputs_.push_back(ToIR(input_proto));
+  }
+  for (const auto& attribute_it : operation_proto.attributes().attributes()) {
+    operation.attributes_.insert(
+        {attribute_it.first, ToIR(attribute_it.second)});
+  }
+}
+
+Value ProtoToIR::ToIR(const proto::Value& value_proto) {
+  std::optional<Value> value;
+  if (value_proto.has_any_value()) {
+    value = Value(ToIR(value_proto.any_value()));
+  } else if (value_proto.has_operation_result_value()) {
+    value = Value(ToIR(value_proto.operation_result_value()));
+  } else if (value_proto.has_block_argument_value()) {
+    value = Value(ToIR(value_proto.block_argument_value()));
+  }
+  CHECK(value) << "Value not constructed due to unexpected variant of Value from IR.";
+  // Register the value with 'SsaNames'.
+  ssa_names_.GetOrCreateID(*value);
+  return *value;
+}
+
+value::Any ProtoToIR::ToIR(const proto::AnyValue&) {
+  value::Any any_proto;
+  // Any is empty.
+  return any_proto;
+}
+
+value::BlockArgument ProtoToIR::ToIR(
+    const proto::BlockArgumentValue& block_argument_proto) {
+  const auto id = block_argument_proto.block_id();
+  auto block_it = blocks_.find(id);
+  CHECK(block_it != blocks_.end()) << "Could not find block with id " << id;
+  Block* block = block_it->second.GetBlockPtr();
+  return value::BlockArgument(*block,
+                              block_argument_proto.block_argument_name());
+}
+
+value::OperationResult ProtoToIR::ToIR(
+    const proto::OperationResultValue& operation_argument_proto) {
+  const auto id = operation_argument_proto.operation_id();
```

Comment:
why `const`?

Reply:
There's no valid mutation to do to the id. The other ids are also const as they shouldn't be changed.